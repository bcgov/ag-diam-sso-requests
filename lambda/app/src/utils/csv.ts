import { format } from '@fast-csv/format';

export const convertCSV = (dataset) => {
  const stream = format({ headers: true });
  let data = '';

  return new Promise((resolve, reject) => {
    stream.on('end', () => resolve(data));
    dataset.map((v) => (data += v));
    stream.end();
  });
};
