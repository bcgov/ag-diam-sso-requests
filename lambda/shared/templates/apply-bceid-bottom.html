<strong>* BCeID production integration is not approved and not ready to use yet</strong>
