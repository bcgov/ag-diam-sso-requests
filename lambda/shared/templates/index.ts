import * as fs from 'fs';
import Handlebars = require('handlebars');
import { noop } from 'lodash';
import { models } from '@lambda-shared/sequelize/models/models';
import { EVENTS, EMAILS } from '@lambda-shared/enums';
import bceidProdApproved from './bceid-prod-approved';
import createIntegrationApproved from './create-integration-approved';
import createIntegrationSubmitted from './create-integration-submitted';
import createIntegrationSubmittedBceidNonprodIdim from './create-integration-submitted-bceid-nonprod-idim';
import createIntegrationSubmittedBceidProd from './create-integration-submitted-bceid-prod';
import deleteIntegrationSubmitted from './delete-integration-submitted';
import deleteIntegrationSubmittedBceid from './delete-integration-submitted-bceid';
import requestLimitExceeded from './request-limit-exceeded';
import teamDeleted from './team-deleted';
import teamInvitation from './team-invitation';
import teamMemberDeletedAdmins from './team-member-deleted-admins';
import teamMemberDeletedUserRemoved from './team-member-deleted-user-removed';
import updateIntegrationApproved from './update-integration-approved';
import updateIntegrationSubmitted from './update-integration-submitted';
import createTeamApiAccountSubmitted from './create-team-api-account-approved';
import createTeamApiAccountApproved from './create-team-api-account-approved';
import deleteTeamApiAccountSubmitted from './delete-team-api-account-submitted';

const APP_URL = process.env.APP_URL || 'http://localhost:3000';
const API_URL = process.env.API_URL || 'http://localhost:8080/app';
const APP_ENV = process.env.APP_ENV || 'development';
const footer = fs.readFileSync(__dirname + '/footer.html', 'utf8');

Handlebars.registerPartial('footer', footer);

const getBuilder = (key: string) => {
  let builder = { render: noop, send: noop };

  switch (key) {
    case EMAILS.BCEID_PROD_APPROVED:
      builder = bceidProdApproved;
      break;
    case EMAILS.CREATE_INTEGRATION_APPROVED:
      builder = createIntegrationApproved;
      break;
    case EMAILS.CREATE_INTEGRATION_SUBMITTED:
      builder = createIntegrationSubmitted;
      break;
    case EMAILS.CREATE_INTEGRATION_SUBMITTED_BCEID_NONPROD_IDIM:
      builder = createIntegrationSubmittedBceidNonprodIdim;
      break;
    case EMAILS.CREATE_INTEGRATION_SUBMITTED_BCEID_PROD:
      builder = createIntegrationSubmittedBceidProd;
      break;
    case EMAILS.DELETE_INTEGRATION_SUBMITTED:
      builder = deleteIntegrationSubmitted;
      break;
    case EMAILS.DELETE_INTEGRATION_SUBMITTED_BCEID:
      builder = deleteIntegrationSubmittedBceid;
      break;
    case EMAILS.TEAM_DELETED:
      builder = teamDeleted;
      break;
    case EMAILS.REQUEST_LIMIT_EXCEEDED:
      builder = requestLimitExceeded;
      break;
    case EMAILS.TEAM_INVITATION:
      builder = teamInvitation;
      break;
    case EMAILS.TEAM_MEMBER_DELETED_ADMINS:
      builder = teamMemberDeletedAdmins;
      break;
    case EMAILS.TEAM_MEMBER_DELETED_USER_REMOVED:
      builder = teamMemberDeletedUserRemoved;
      break;
    case EMAILS.UPDATE_INTEGRATION_APPROVED:
      builder = updateIntegrationApproved;
      break;
    case EMAILS.UPDATE_INTEGRATION_SUBMITTED:
      builder = updateIntegrationSubmitted;
      break;
    case EMAILS.CREATE_TEAM_API_ACCOUNT_SUBMITTED:
      builder = createTeamApiAccountSubmitted;
      break;
    case EMAILS.CREATE_TEAM_API_ACCOUNT_APPROVED:
      builder = createTeamApiAccountApproved;
      break;
    case EMAILS.DELETE_TEAM_API_ACCOUNT_SUBMITTED:
      builder = deleteTeamApiAccountSubmitted;
      break;
    default:
      break;
  }

  return builder;
};

export interface RenderResult {
  code?: string;
  subject: string;
  body: string;
}

export const renderTemplate = async (code: string, data: any): Promise<RenderResult> => {
  data.appUrl = APP_URL;
  data.apiUrl = API_URL;
  const prefix = APP_ENV === 'production' ? '' : '[DEV] ';
  const builder = getBuilder(code);
  const result = await builder.render(data);

  if (prefix) result.subject = `${prefix}${result.subject}`;
  result.code = code;
  return result;
};

const createEvent = async (data) => {
  try {
    await models.event.create(data);
  } catch (err) {
    console.error(err);
  }
};

export const sendTemplate = async (code: string, data: any) => {
  try {
    const builder = getBuilder(code);
    const rendered = await renderTemplate(code, data);
    await builder.send(data, rendered);
  } catch (err) {
    console.error(err);

    if (data.integration) {
      createEvent({
        eventCode: EVENTS.EMAIL_SUBMISSION_FAILURE,
        requestId: data.integration.id,
        details: { emailCode: code, error: err.message || err },
      });
    }
  }
};

export const sendTemplates = async (emails: { code: string; data: any }[]) => {
  await emails.map((email) => sendTemplate(email.code, email.data));
};

export default { renderTemplate, sendTemplate };
