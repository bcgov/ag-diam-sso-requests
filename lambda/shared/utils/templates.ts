const APP_URL = process.env.APP_URL || '';
const APP_ENV = process.env.APP_ENV || 'development';

type EmailMessage =
  | 'create-request-submitted'
  | 'create-request-approved'
  | 'uri-change-request-submitted'
  | 'uri-change-request-approved'
  | 'request-deleted'
  | 'request-deleted-notification-to-admin'
  | 'request-limit-exceeded';

interface BodyData {
  projectName?: string;
  requestNumber?: number;
  requestOwner?: string;
  submittedBy?: string;
}

const footer = `
<p>
    If you have any questions, please contact us by
    <a href="https://chat.developer.gov.bc.ca/channel/sso" target="_blank" title="Rocket Chat" rel="noreferrer">
    Rocket.Chat
    </a>
    or email at:
    <a href="mailto:zorin.samji@gov.bc.ca" title="Pathfinder SSO" target="_blank" rel="noreferrer">
        Zorin.Samji@gov.bc.ca
    </a>
</p>
<p>Thank you.<br />Pathfinder SSO Team</p>`;

export const getEmailBody = (
  messageType: EmailMessage,
  { projectName = '', requestNumber = -1, requestOwner = '', submittedBy = '' }: BodyData,
) => {
  switch (messageType) {
    case 'create-request-submitted':
      return `
        <h1>Your Pathfinder SSO request #${requestNumber} is successfully submitted.</h1>
        <p>
            <strong>Project name: </strong>${projectName}<br /><strong>Submitted by:</strong>${submittedBy}
        </p>
        <p>
            <strong>The expected processing time is 20 minutes</strong><br />
            Once the request is approved, you will receive another email once your JSON Client Installation is ready. If you are
            not the requester, this email serves only to notify you of the request status.
        </p>
        ${footer}`;

    case 'create-request-approved':
      return `
        <h1>Your Pathfinder SSO request #${requestNumber} is approved.</h1>
        <p>
            <strong>Project name: </strong>${projectName}<br /><strong>Submitted by:</strong>${submittedBy}
        </p>
        <p>
            ${submittedBy}, please <a href="${APP_URL}/my-requests" title="Log in" target="_blank" rel="noreferrer">Log in</a>
             to your dashboard to access JSON Client Installation File.
            If you are not the requester, this email serves only to notify you of the request status.
        </p>
        ${footer}`;

    case 'uri-change-request-submitted':
      return `
        <h1>Your Pathfinder SSO URI change request #${requestNumber} is successfully submitted.</h1>
        <p>
            <strong>Project name: </strong>${projectName}<br /><strong>Submitted by:</strong>${submittedBy}
        </p>
        <p>
            <strong>The expected processing time is 20 minutes</strong><br />
            Once the request is approved, you will receive another email once your JSON Client Installation is ready. If you are
            not the requester, this email serves only to notify you of the request status.
        </p>
        ${footer}`;

    case 'uri-change-request-approved':
      return `
        <h1>Your Pathfinder SSO URI change request #${requestNumber} is approved.</h1>
        <p>
            <strong>Project name: </strong>${projectName}<br /><strong>Submitted by:</strong>${submittedBy}
        </p>
        <p>
            ${submittedBy}, please <a href="${APP_URL}/my-requests" title="Log in" target="_blank" rel="noreferrer">Log in</a>
             to your dashboard to access JSON Client Installation File.
            If you are not the requester, this email serves only to notify you of the request status.
        </p>
        ${footer}`;

    case 'request-deleted':
      return `
        <h1>Your Pathfinder SSO request #${requestNumber} is deleted.</h1>
        <p>
            <strong>Project name: </strong>${projectName}<br /><strong>Deleted by:</strong>${submittedBy}
        </p>
        <p>This email serves only to notify you of the request status.</p>
        ${footer}`;

    case 'request-deleted-notification-to-admin':
      return `
        <h1>Pathfinder SSO request #${requestNumber} is deleted by ${submittedBy}</h1>
        <p>
            <strong>Project name: </strong>${projectName}
        </p>`;
    case 'request-limit-exceeded':
      return `
        <h1>Request Limit Exceeded</h1>
        <p>Pathfinder SSO user ${submittedBy} has exceeded their daily limit</p>
        `;

    default:
      return '';
  }
};

export const getEmailSubject = (messageType: EmailMessage) => {
  const prefix = APP_ENV === 'development' ? '[DEV] ' : '';

  switch (messageType) {
    case 'create-request-submitted':
      return `${prefix}Pathfinder SSO request submitted`;
    case 'create-request-approved':
      return `${prefix}Pathfinder SSO request approved`;
    case 'uri-change-request-submitted':
      return `${prefix}Pathfinder SSO URI change request submitted`;
    case 'uri-change-request-approved':
      return `${prefix}Pathfinder SSO URI change request approved`;
    case 'request-deleted':
      return `${prefix}Pathfinder SSO request deleted`;
    case 'request-deleted-notification-to-admin':
      return `${prefix}Pathfinder SSO request deleted`;
    case 'request-limit-exceeded':
      return `${prefix}Pathfinder SSO request limit reached`;
    default:
      return '';
  }
};
