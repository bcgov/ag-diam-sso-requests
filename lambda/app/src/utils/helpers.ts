import { Request } from '../interfaces/Request';
import validate from 'react-jsonschema-form/lib/validate';
import requesterSchema from '../schemas/requester-info';
import providerSchema from '../schemas/providers';
import termsAndConditionsSchema from '../schemas/terms-and-conditions';
import { isObject, omit, sortBy } from 'lodash';
import { customValidate } from './customValidate';
import { diff } from 'deep-diff';
import { Session, Data } from '../../../shared/interfaces';
import { Op } from 'sequelize';
import { getEmailBody, getEmailSubject, EmailMessage } from '../../../shared/utils/templates';
import { sendEmail } from '../../../shared/utils/ches';

export const errorMessage = 'No changes submitted. Please change your details to update your integration.';

export const omitNonFormFields = (data: Request) =>
  omit(data, [
    'updatedAt',
    'createdAt',
    'archived',
    'status',
    'environments',
    'actionNumber',
    'prNumber',
    'clientName',
    'idirUserid',
    'idirUserDisplayName',
    'id',
  ]);

export type BceidEvent = 'submission' | 'deletion' | 'update';
const bceidRealms = ['onestopauth-basic', 'onestopauth-business', 'onestopauth-both'];
export const usesBceid = (realm: string | undefined) => {
  if (!realm) return false;
  return bceidRealms.includes(realm);
};

export const notifyIdim = async (request: Data, bceidEvent: BceidEvent) => {
  const { realm, id, preferredEmail, additionalEmails, environments } = request;
  if (!usesBceid(realm) || bceidEvent === 'update') return;
  const usesProd = environments.includes('prod');

  // Only cc user for production requests
  let cc = usesProd ? [preferredEmail] : [];
  if (Array.isArray(additionalEmails) && usesProd) cc = cc.concat(additionalEmails);

  let emailCode: EmailMessage;
  if (bceidEvent === 'submission' && usesProd) emailCode = 'bceid-idim-prod-submitted';
  else if (bceidEvent === 'submission') emailCode = 'bceid-idim-prod-submitted';
  else if (bceidEvent === 'deletion') emailCode = 'bceid-idim-deleted';
  else return;
  // const to = APP_ENV === 'production' ? ['bcgov.sso@gov.bc.ca', 'IDIM.Consulting@gov.bc.ca'] : ['bcgov.sso@gov.bc.ca'];
  const to = ['bcgov.sso@gov.bc.ca'];
  return sendEmail({
    to,
    cc,
    body: getEmailBody(emailCode, request),
    subject: getEmailSubject(emailCode, id),
    event: { emailCode, requestId: id },
  });
};

const sortURIFields = (data: any) => {
  const sortedData = { ...data };
  const { devValidRedirectUris, testValidRedirectUris, prodValidRedirectUris } = data;
  sortedData.devValidRedirectUris = sortBy(devValidRedirectUris);
  sortedData.testValidRedirectUris = sortBy(testValidRedirectUris);
  sortedData.prodValidRedirectUris = sortBy(prodValidRedirectUris);
  return sortedData;
};

export const processRequest = (data: any) => {
  const immutableFields = ['idirUserid', 'projectLead', 'clientName', 'status'];
  const allowedRequest = omit(data, immutableFields);
  const sortedRequest = sortURIFields(allowedRequest);
  sortedRequest.environments = processEnvironments(sortedRequest);
  return sortedRequest;
};

const processEnvironments = (data: any) => {
  const environments = [];
  if (data.dev) environments.push('dev');
  if (data.test) environments.push('test');
  if (data.prod) environments.push('prod');
  return environments;
};

export const getDifferences = (newData: any, originalData: Request) => {
  const sortedNewData = sortURIFields(newData);
  return diff(omitNonFormFields(originalData), omitNonFormFields(sortedNewData));
};

export const validateRequest = (formData: any, original: Request, isUpdate = false) => {
  if (isUpdate) {
    const differences = getDifferences(formData, original);
    if (!differences) return { message: errorMessage };
  }

  const { errors: firstPageErrors } = validate(formData, requesterSchema);
  const { errors: secondPageErrors } = validate(formData, providerSchema, customValidate);
  const { errors: thirdPageErrors } = validate(formData, termsAndConditionsSchema);
  const allValid = firstPageErrors.length === 0 && secondPageErrors.length === 0 && thirdPageErrors.length === 0;
  if (allValid) return true;
  return {
    firstPageErrors,
    secondPageErrors,
    thirdPageErrors,
  };
};

// GH actions inputs expects an object where all values are strings
export const stringifyGithubInputs = (inputs: any) => {
  const stringifiedInputs = {};
  Object.entries(inputs).forEach(([key, value]) => {
    if (isObject(value) || Array.isArray(value)) {
      stringifiedInputs[key] = JSON.stringify(value);
    } else {
      stringifiedInputs[key] = String(value);
    }
  });

  return stringifiedInputs;
};

export const isAdmin = (session: Session) => session.client_roles?.includes('sso-admin');

export const getWhereClauseForAllRequests = (data: {
  searchField: string[];
  searchKey: string;
  status?: string;
  archiveStatus?: string;
  realms?: string[];
  environments?: string[];
}) => {
  const where: any = {};
  const { searchField, searchKey, status = [], archiveStatus = [], realms, environments } = data;

  if (searchKey && searchField && searchField.length > 0) {
    where[Op.or] = [];
    searchField.forEach((field) => {
      if (field === 'id') {
        const id = Number(searchKey);
        if (!Number.isNaN(id)) where[Op.or].push({ id });
      } else {
        where[Op.or].push({ [field]: { [Op.iLike]: `%${searchKey}%` } });
      }
    });
  }

  if (status.length > 0) {
    where.status = {
      [Op.in]: status,
    };
  }

  if (archiveStatus.length === 1) {
    where.archived = archiveStatus[0] === 'archived';
  }

  if (realms)
    where.realm = {
      [Op.in]: realms,
    };

  if (environments)
    where.environments = {
      [Op.overlap]: environments,
    };

  return where;
};
