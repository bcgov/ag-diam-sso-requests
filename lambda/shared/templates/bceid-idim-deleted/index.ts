import fs from 'fs';
import Handlebars from 'handlebars';
import { processRequest } from '../helpers';

const SUBJECT_TEMPLATE = `BCeID Integration Request ID {{request.id}} Deleted`;
const template = fs.readFileSync(__dirname + '/template.html', 'utf8');

const subjectHandler = Handlebars.compile(SUBJECT_TEMPLATE, { noEscape: true });
const bodyHandler = Handlebars.compile(template, { noEscape: true });

export const render = (originalData) => {
  const { request, appUrl } = originalData;
  const data = { request: processRequest(request), appUrl };
  return {
    subject: subjectHandler(data),
    body: bodyHandler(data),
  };
};

export default { render };
