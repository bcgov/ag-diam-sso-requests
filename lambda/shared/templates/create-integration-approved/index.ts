import * as fs from 'fs';
import Handlebars = require('handlebars');
import { processRequest } from '../helpers';
import { Data } from '@lambda-shared/interfaces';
import { sendEmail } from '@lambda-shared/utils/ches';
import { SSO_EMAIL_ADDRESS } from '@lambda-shared/local';
import { getIntegrationEmails } from '../helpers';
import type { RenderResult } from '../index';

const SUBJECT_TEMPLATE = `Pathfinder SSO request approved`;
const template = fs.readFileSync(__dirname + '/template.html', 'utf8');

const subjectHandler = Handlebars.compile(SUBJECT_TEMPLATE, { noEscape: true });
const bodyHandler = Handlebars.compile(template, { noEscape: true });

interface DataProps {
  integration: Data;
}

export const render = async (originalData: DataProps): Promise<RenderResult> => {
  const { integration } = originalData;
  const data = { ...originalData, integration: await processRequest(integration) };

  return {
    subject: subjectHandler(data),
    body: bodyHandler(data),
  };
};

export const send = async (data: DataProps, rendered: RenderResult) => {
  const { integration } = data;
  const emails = await getIntegrationEmails(integration);

  return sendEmail({
    to: emails,
    cc: [SSO_EMAIL_ADDRESS],
    ...rendered,
  });
};

export default { render, send };
