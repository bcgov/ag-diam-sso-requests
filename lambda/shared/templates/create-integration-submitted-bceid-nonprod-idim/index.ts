import * as fs from 'fs';
import Handlebars = require('handlebars');
import { processRequest } from '../helpers';
import { Data } from '@lambda-shared/interfaces';
import { sendEmail } from '@lambda-shared/utils/ches';
import { SSO_EMAIL_ADDRESS, IDIM_EMAIL_ADDRESS } from '@lambda-shared/local';

const SUBJECT_TEMPLATE = `New BCeID Request ID {{integration.id}}`;
const template = fs.readFileSync(__dirname + '/template.html', 'utf8');

const subjectHandler = Handlebars.compile(SUBJECT_TEMPLATE, { noEscape: true });
const bodyHandler = Handlebars.compile(template, { noEscape: true });

interface DataProps {
  integration: Data;
  appUrl: string;
}

export const render = (originalData: DataProps) => {
  const { integration, appUrl } = originalData;
  const data = { integration: processRequest(integration), appUrl };

  return {
    subject: subjectHandler(data),
    body: bodyHandler(data),
  };
};

export const send = async (data: DataProps) => {
  return sendEmail({
    to: [IDIM_EMAIL_ADDRESS],
    cc: [SSO_EMAIL_ADDRESS],
    ...render(data),
  });
};

export default { render, send };
