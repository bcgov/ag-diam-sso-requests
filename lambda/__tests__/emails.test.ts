import { getEmailBody } from '../shared/utils/templates';
import { formDataDev, formDataProd } from './fixtures';

describe('Email snapshots', () => {
  it('Should return the expeted email for created requests', () => {
    expect(getEmailBody('create-request-submitted', formDataDev)).toMatchSnapshot();
  });
  it('Should return the expeted email for approved', () => {
    expect(getEmailBody('create-request-approved', formDataDev)).toMatchSnapshot();
  });
  it('Should return the expeted email for updated submissions', () => {
    expect(getEmailBody('uri-change-request-submitted', formDataDev)).toMatchSnapshot();
  });
  it('Should return the expeted email for updated approvals', () => {
    expect(getEmailBody('uri-change-request-approved', formDataDev)).toMatchSnapshot();
  });
  it('Should return the expeted email for deletions', () => {
    expect(getEmailBody('request-deleted', formDataDev)).toMatchSnapshot();
  });
  it('Should return the expeted email for deletions to admins', () => {
    expect(getEmailBody('request-deleted-notification-to-admin', formDataDev)).toMatchSnapshot();
  });
  it('Should return the expeted email for exceeding the limit', () => {
    expect(getEmailBody('request-limit-exceeded', formDataDev)).toMatchSnapshot();
  });
  it('Should return the expeted email for Dev IDIM submissions', () => {
    expect(getEmailBody('bceid-idim-dev-submitted', formDataDev)).toMatchSnapshot();
  });
  it('Should return the expeted email for Prod IDIM submissions', () => {
    expect(getEmailBody('bceid-idim-prod-submitted', formDataProd)).toMatchSnapshot();
  });
  it('Should return the expeted email for IDIM deletions', () => {
    expect(getEmailBody('bceid-idim-deleted', formDataProd)).toMatchSnapshot();
  });
  it('Should return the expeted email for bceid approvals', () => {
    expect(getEmailBody('bceid-request-approved', formDataDev)).toMatchSnapshot();
  });
});
