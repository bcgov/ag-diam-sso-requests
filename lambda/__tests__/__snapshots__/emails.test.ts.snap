// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Email snapshots Should return the expeted email for Dev IDIM submissions 1`] = `
"
      <h1>Hello IDIM Team,</h1>
      <p>
        We are notifying you that a new dev and/or test integration request has been submitted. The request details are below:
      </p>

      <ul>
        <li><strong>Project name:</strong> test</li>
        <li><strong>Accountable person:</strong> test user</li>
        <li><strong>URIs:</strong>
          <ul>
            <li> Development: https://b </li>
            
          </ul>
        </li>
        <li><strong>Identity Providers Required:</strong> idir</li>
      </ul>

      <p>Thank you,</p>

      <p>Pathfinder SSO team.</p>
      "
`;

exports[`Email snapshots Should return the expeted email for IDIM deletions 1`] = `
"
      <h1>Hello IDIM Team,</h1>
      <p>
        We are notifying you that the client for the following integration request has requested a deletion. The request details are below:
      </p>

      <ul>
        <li><strong>Project name:</strong> test</li>
        <li><strong>Accountable person:</strong> test user</li>
        <li><strong>URIs:</strong>
          <ul>
            <li> Development: https://b </li>
            <li> Test: https://a </li>
            <li> Production: https://a </li>
          </ul>
        </li>
        <li><strong>Identity Providers Required:</strong> idir</li>
      </ul>

      <p>Thank you,</p>

      <p>Pathfinder SSO team.</p>
      "
`;

exports[`Email snapshots Should return the expeted email for approved 1`] = `
"
        <h1>Your Pathfinder SSO request #1 is approved.</h1>
        <p>
            <strong>Project name: </strong>test<br /><strong>Submitted by:</strong>test user
        </p>
        <p>
            test user, please <a href=\\"/my-requests\\" title=\\"Log in\\" target=\\"_blank\\" rel=\\"noreferrer\\">Log in</a>
             to your dashboard to access JSON Client Installation File.
            If you are not the requester, this email serves only to notify you of the request status.
        </p>
        
<p>
    If you have any questions, please contact us by
    <a href=\\"https://chat.developer.gov.bc.ca/channel/sso\\" target=\\"_blank\\" title=\\"Rocket Chat\\" rel=\\"noreferrer\\">
    Rocket.Chat
    </a>
    or email at:
    <a href=\\"mailto:bcgov.sso@gov.bc.ca\\" title=\\"Pathfinder SSO\\" target=\\"_blank\\" rel=\\"noreferrer\\">
    bcgov.sso@gov.bc.ca
    </a>
</p>
<p>Thank you.<br />Pathfinder SSO Team</p>"
`;

exports[`Email snapshots Should return the expeted email for bceid approvals 1`] = `
"
        <h1>Hello Pathfinder Friend,</h1>
        <p>Your Pathfinder SSO production environment for BCeID <strong>integration request ID 1</strong> has been approved.</p>
        <p>Your resources will be available in the next 20 minutes or so, please log in to your dashboard to see the progress.</p>
        <p><strong>Project Name: </strong> test</p>
        <p><strong>Submitted By: </strong> test user</p>

        <p>If you are not the requester, this email serves only to notify you of the request status.</p>
        
<p>
    If you have any questions, please contact us by
    <a href=\\"https://chat.developer.gov.bc.ca/channel/sso\\" target=\\"_blank\\" title=\\"Rocket Chat\\" rel=\\"noreferrer\\">
    Rocket.Chat
    </a>
    or email at:
    <a href=\\"mailto:bcgov.sso@gov.bc.ca\\" title=\\"Pathfinder SSO\\" target=\\"_blank\\" rel=\\"noreferrer\\">
    bcgov.sso@gov.bc.ca
    </a>
</p>
<p>Thank you.<br />Pathfinder SSO Team</p>
        "
`;

exports[`Email snapshots Should return the expeted email for created requests 1`] = `
"
        <h1>Your Pathfinder SSO request #1 is successfully submitted.</h1>
        <p>
            <strong>Project name: </strong>test<br /><strong>Submitted by: </strong>test user
        </p>
        <p>
            <strong>The expected processing time is 20 minutes</strong><br />
            Once the request is approved, you will receive another email that your JSON Client Installation is ready. If you are
            not the requester, this email serves only to notify you of the request status.
        </p>
        
<p>
    If you have any questions, please contact us by
    <a href=\\"https://chat.developer.gov.bc.ca/channel/sso\\" target=\\"_blank\\" title=\\"Rocket Chat\\" rel=\\"noreferrer\\">
    Rocket.Chat
    </a>
    or email at:
    <a href=\\"mailto:bcgov.sso@gov.bc.ca\\" title=\\"Pathfinder SSO\\" target=\\"_blank\\" rel=\\"noreferrer\\">
    bcgov.sso@gov.bc.ca
    </a>
</p>
<p>Thank you.<br />Pathfinder SSO Team</p>"
`;

exports[`Email snapshots Should return the expeted email for deletions 1`] = `
"
        <h1>Your Pathfinder SSO request #1 is deleted.</h1>
        <p>
            <strong>Project name: </strong>test<br /><strong>Deleted by:</strong>test user
        </p>
        <p>This email serves only to notify you of the request status.</p>
        
<p>
    If you have any questions, please contact us by
    <a href=\\"https://chat.developer.gov.bc.ca/channel/sso\\" target=\\"_blank\\" title=\\"Rocket Chat\\" rel=\\"noreferrer\\">
    Rocket.Chat
    </a>
    or email at:
    <a href=\\"mailto:bcgov.sso@gov.bc.ca\\" title=\\"Pathfinder SSO\\" target=\\"_blank\\" rel=\\"noreferrer\\">
    bcgov.sso@gov.bc.ca
    </a>
</p>
<p>Thank you.<br />Pathfinder SSO Team</p>"
`;

exports[`Email snapshots Should return the expeted email for deletions to admins 1`] = `
"
        <h1>Pathfinder SSO request #1 is deleted by test user</h1>
        <p>
            <strong>Project name: </strong>test
        </p>"
`;

exports[`Email snapshots Should return the expeted email for exceeding the limit 1`] = `
"
        <h1>Request Limit Exceeded</h1>
        <p>Pathfinder SSO user test user has exceeded their daily limit</p>
        "
`;

exports[`Email snapshots Should return the expeted email for updated approvals 1`] = `
"
        <h1>Your Pathfinder SSO change for request #1 is approved.</h1>
        <p>
            <strong>Project name: </strong>test<br /><strong>Submitted by:</strong>test user
        </p>
        <p>
            test user, please <a href=\\"/my-requests\\" title=\\"Log in\\" target=\\"_blank\\" rel=\\"noreferrer\\">Log in</a>
             to your dashboard to access JSON Client Installation File.
            If you are not the requester, this email serves only to notify you of the request status.
        </p>
        
<p>
    If you have any questions, please contact us by
    <a href=\\"https://chat.developer.gov.bc.ca/channel/sso\\" target=\\"_blank\\" title=\\"Rocket Chat\\" rel=\\"noreferrer\\">
    Rocket.Chat
    </a>
    or email at:
    <a href=\\"mailto:bcgov.sso@gov.bc.ca\\" title=\\"Pathfinder SSO\\" target=\\"_blank\\" rel=\\"noreferrer\\">
    bcgov.sso@gov.bc.ca
    </a>
</p>
<p>Thank you.<br />Pathfinder SSO Team</p>"
`;

exports[`Email snapshots Should return the expeted email for updated submissions 1`] = `
"
        <h1>Your Pathfinder SSO change for request #1 is successfully submitted.</h1>
        <p>
            <strong>Project name: </strong>test<br /><strong>Submitted by:</strong>test user
        </p>
        <p>
            <strong>The expected processing time is 20 minutes</strong><br />
            Once the request is approved, you will receive another email once your JSON Client Installation is ready. If you are
            not the requester, this email serves only to notify you of the request status.
        </p>
        
<p>
    If you have any questions, please contact us by
    <a href=\\"https://chat.developer.gov.bc.ca/channel/sso\\" target=\\"_blank\\" title=\\"Rocket Chat\\" rel=\\"noreferrer\\">
    Rocket.Chat
    </a>
    or email at:
    <a href=\\"mailto:bcgov.sso@gov.bc.ca\\" title=\\"Pathfinder SSO\\" target=\\"_blank\\" rel=\\"noreferrer\\">
    bcgov.sso@gov.bc.ca
    </a>
</p>
<p>Thank you.<br />Pathfinder SSO Team</p>"
`;
